Merge train: infer candidate from state

rudymatela opened the pull request:
... in preparation for #77 (partly closes: #77)

The changes in this PR do not change Hoff behaviour, the high level functionality should be the same.

Instead of maintaining a global candidate, we infer it from the state always.  This is in preparation for merge trains (#77), where we will have multiple builds running in parallel and thus multiple candidates.

* The `ProjectState` datatype is now simpler and easier to maintain (before we could risk it being inconsistent);
* Some code paths are now simpler, as we don't have to do the work maintaining the global candidate;
* There is a new explicit integration status `Promoted` to signify that the PR branch is now part of `master`, before this was signified by being `Integrated ... BuildSucceeded` but not being the global candidate.

A linear search on the number of open PRs is imposed.  I believe this is not a problem because:

* the number of PRs should be at most a few dozen, examples:
    - there are only [5 open PRs on channable/hoff](https://github.com/channable/hoff/pulls) currently;
    - a dozen open PRs seems to be a reasonable average (:wink:);
    - a busy project (:wink:) should have between 100 and 200 open PRs in which a linear search is still reasonable;
    - ... and we could still have way more than that without suffering significant performance impact (1000 or 10000 __open__ PRs).
* we already perform [linear search](https://github.com/channable/hoff/blob/0c3e53b1c1ec2fc8b4e1d68d5dc02fb278a65d1e/src/Project.hs#L298-L315) for [other operations](https://github.com/channable/hoff/blob/0c3e53b1c1ec2fc8b4e1d68d5dc02fb278a65d1e/src/Project.hs#L366-L369), the candidate was just a special case in where we don't.  If we were to actually get rid of linear search altogether, we perhaps should have a more generic solution that covers all cases.  (If linear search were to be a problem, we would already be suffering from it.)
* with #77, we are bound to perform a linear search on the candidate PRs anyway.

### Notes

This is an alternative to PR #130.  Here, instead of changing the datatype of `integrationCandidate`, we simply remove it and try to infer from the rest of the state every time.

rudymatela commented:
c'mon GitHub, I thought you could handle the renaming of branches... apparently not.

fatho commented on tests/Spec.hs:
```diff
@@ -1353,16 +1344,16 @@ main = hspec $ do
       cId     `shouldBe` PullRequestId 2
       actions `shouldBe`
-        [ ATryIntegrate "Merge #2: Add my test results\n\nApproved-by: deckard\nAuto-deploy: false\n" (Branch "refs/pull/2/head", Sha "f37") False
+        [ ATryPromote (Branch "results/leon") (Sha "38d")
```
Where does this extra action come from?

rudymatela replied:
In this test suite, there are two open PRs:

* the first which has been rebased and the build has (just) passed;
* the second which was not rebased yet;

So the actions should be to promote the first to be the new master, and try to rebase (integrate) the second.

If we instead set the first PR to be `Promoted`, the `TryPromote` should disappear.

This extra action just came to how the state is represented now: before a `Promoted` pr was one that was `Integrated BuildSuceeded` while the candidate was `Nothing`.

I decided to change the test to include the promotion.

rudymatela replied:
I could update the test name, but I think it still applies :grin::

> "picks a new candidate from the queue after a successful push"

fatho reviewed (commented):
Basically LGTM, but some questions:

rudymatela commented on src/Project.hs:
```diff
@@ -287,6 +295,7 @@ classifyPullRequest pr = case approval pr of
       BuildPending    -> PrStatusBuildPending
       BuildSucceeded  -> PrStatusIntegrated
       BuildFailed url -> PrStatusFailedBuild url
+    Promoted -> PrStatusIntegrated -- TODO: state-of-its-own?
```
Not really, I removed it from the code now.

rudymatela commented:
@fatho Thanks for the review.

Since I think I have addressed your comments, I am going to merge and tag now.

I will leave the deploy for later today at around 19:00 (if there are no items in the build queue).

@OpsBotPrime merge and tag

OpsBotPrime commented:
Pull request approved for merge and tag by @rudymatela, rebasing now.

OpsBotPrime commented:
@rudymatela Sorry, I could not tag your PR. The previous tag `v0.25.6` seems invalid

rudymatela commented:
> @rudymatela Sorry, I could not tag your PR. The previous tag `v0.25.6` seems invalid

Oohh yes, of course.  #117 will make you support it.
